rust/sbd-server/src/cslot.rs

```diff
@@ -0,0 +1,416 @@
+        let (uniq, index, ws) = self.get_sender(pk)?;
+
+        match ws.send(payload).await {
+            Err(err) => {
+                self.remove(uniq, index);
+                Err(err)
+            }
+            Ok(_) => Ok(()),
+        }
+    }
+}
+
+async fn top_task(
+    config: Arc<Config>,
+    ip_rate: Arc<ip_rate::IpRate>,
+    weak: WeakCSlot,
+    mut recv: tokio::sync::mpsc::UnboundedReceiver<TaskMsg>,
+) {
+    let mut item = recv.recv().await;
+    loop {
+        let uitem = match item {
+            None => break,
+            Some(uitem) => uitem,
+        };
+
+        item = if let TaskMsg::NewWs {
+            uniq,
+            index,
+            ws,
+            ip,
+            pk,
+        } = uitem
+        {
+            let next_i = tokio::select! {
+                i = recv.recv() => Some(i),
+                _ = ws_task(
+                    &config,
+                    &ip_rate,
+                    &weak,
+                    &ws,
+                    ip,
+                    pk,
+                    uniq,
+                    index,
+                ) => None,
+            };
+
+            ws.close().await;
+            drop(ws);
+            if let Some(cslot) = weak.upgrade() {
+                cslot.remove(uniq, index);
+            }
+
+            match next_i {
+                Some(i) => i,
+                None => recv.recv().await,
+            }
+        } else {
+            recv.recv().await
+        };
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+async fn ws_task(
+    config: &Arc<Config>,
+    ip_rate: &ip_rate::IpRate,
+    weak_cslot: &WeakCSlot,
+    ws: &Arc<ws::WebSocket<MaybeTlsStream>>,
+    ip: Arc<std::net::Ipv6Addr>,
+    pk: PubKey,
+    uniq: u64,
+    index: usize,
+) {
+    let auth_res = tokio::time::timeout(config.idle_dur(), async {
+        use rand::Rng;
+        let mut nonce = [0xdb; 32];
+        rand::thread_rng().fill(&mut nonce[..]);
+
+        ws.send(cmd::SbdCmd::auth_req(&nonce)).await?;
+
+        let auth_res = ws.recv().await?;
+
+        if !ip_rate.is_ok(&ip, auth_res.as_ref().len()).await {
+            return Err(Error::other("ip rate limited"));
+        }
+
+        if let cmd::SbdCmd::AuthRes(sig) = cmd::SbdCmd::parse(auth_res)? {
+            if !pk.verify(&sig, &nonce) {
+                return Err(Error::other("invalid sig"));
+            }
+        } else {
+            return Err(Error::other("invalid auth response"));
+        }
+
+        // NOTE: the byte_nanos limit is sent during the cslot insert
+
+        ws.send(cmd::SbdCmd::limit_idle_millis(config.limit_idle_millis))
+            .await?;
+
+        if let Some(cslot) = weak_cslot.upgrade() {
+            cslot.mark_ready(uniq, index);
+        } else {
+            return Err(Error::other("closed"));
+        }
+
+        ws.send(cmd::SbdCmd::ready()).await?;
+
+        Ok(())
+    })
+    .await;
+
+    if auth_res.is_err() {
+        return;
```

Comment:
would it be helpful to `eprintln` this for debugging?

Reply:
I'm experimenting with not having tracing enabled on the server. There is overhead to tracing that I'd like to avoid if possible. Separately, we may someday want some kind of audit log that could show when clients connect and when/why they are disconnected. This data could make it into that if we choose to implement that.